plutus-tx-plugin/test/Plugin/Spec.hs

```diff
@@ -387,3 +415,16 @@ valueRestriction = plc @"valueRestriction" (let { f :: forall a . a; f = Builtin
 
 recordSelector :: CompiledCode (MyMonoRecord -> Int)
 recordSelector = plc @"recordSelector" (\(x :: MyMonoRecord) -> mrA x)
+
+emptyRoseId1 :: CompiledCode (EmptyRose -> EmptyRose)
+emptyRoseId1 = plc @"emptyRoseId1" (
+    let map _ []     = []
+        map f (x:xs) = f x : map f xs
+        go (EmptyRose xs) = EmptyRose (map go xs)
+    in go)
```

Comment:
This gives "Violation of the value restriction: Type abstraction body is not a value".

Reply:
The problem is that GHC turns this into `/\ a b. let map = ....`, rather than `let map = /\ a b. ...`. We can't translate the former straightforwardly into Plutus Core, because a let translates to an immediately applied lambda, i.e. not a value, so we hit the value restriction.

IMO this is just more evidence that the value restriction is a bad idea, but you can work around it by making sure your type abstraction happens inside the let, by giving a polymorphic type signature to `map`.

Reply:
> `/\ a b. let map = ....`, rather than `let map = /\ a b. ....`

I guess it's something like `map_poly = /\ a b. let map_mono = ....`? Because we can type check `map_poly go`, but not `map_mono go`, because `go` is not of type `a -> b` where `a` and `b` are free.

Reply:
Must be something like that. I don't know why it does that, but adding the signature you actually want usually fixes it.

Reply:
That's not too user-friendly, perhaps we should indeed bring the value restriction topic again. At least we should document this design choice.